### ISSUE TITLE
[xgboost4j-spark] Occasional SIGSEGV right when starting Rabit

### ISSUE THREAD

**Issue by randomjohnnyh** [dmlc/xgboost#1394]
Sometimes when launching xgboost.train across 100+ workers, I get a SIGSEGV from JNI because of a memcpy.

The problem does not occur when I limit to one ~~worker~~ executor per machine, but I would like to maximize the speed at which I can preprocess the data before feeding into xgboost, so I would like to use one core per executor.

It does not happen consistently, so maybe it is some kind of race condition?

This is the full Java stack trace:
```
Stack: [0x00007fd1d50a7000,0x00007fd1d51a8000],  sp=0x00007fd1d51a5720,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libc.so.6+0x35a35]  exit+0x35

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  ml.dmlc.xgboost4j.java.XGBoostJNI.RabitInit([Ljava/lang/String;)I+0
j  ml.dmlc.xgboost4j.java.Rabit.init(Ljava/util/Map;)V+95
j  ml.dmlc.xgboost4j.scala.spark.XGBoost$$anonfun$buildDistributedBoosters$1.apply(Lscala/collection/Iterator;)Lscala/collection/Iterator;+40
j  ml.dmlc.xgboost4j.scala.spark.XGBoost$$anonfun$buildDistributedBoosters$1.apply(Ljava/lang/Object;)Ljava/lang/Object;+5
j  org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$20.apply(Lorg/apache/spark/TaskContext;ILscala/collection/Iterator;)Lscala/collection/Iterator;+5
j  org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$20.apply(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+13
j  org.apache.spark.rdd.MapPartitionsRDD.compute(Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;)Lscala/collection/Iterator;+27
j  org.apache.spark.rdd.RDD.computeOrReadCheckpoint(Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;)Lscala/collection/Iterator;+26
j  org.apache.spark.CacheManager.getOrCompute(Lorg/apache/spark/rdd/RDD;Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;Lorg/apache/spark/storage/StorageLevel;)Lscala/collection/Iterator;+196
J 6098 C1 org.apache.spark.rdd.RDD.iterator(Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;)Lscala/collection/Iterator; (59 bytes) @ 0x00007fd221e3382c [0x00007fd221e32a00+0xe2c]
j  org.apache.spark.scheduler.ResultTask.runTask(Lorg/apache/spark/TaskContext;)Ljava/lang/Object;+151
j  org.apache.spark.scheduler.Task.run(JILorg/apache/spark/metrics/MetricsSystem;)Lscala/Tuple2;+118
j  org.apache.spark.executor.Executor$TaskRunner.run()V+342
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
```

**Comment by CodingCat** [comment]
when you say `one worker per machine `, you mean `worker` which starts the executors, or you are referring to the executor which runs the tasks?

**Comment by randomjohnnyh** [comment]
Ah, I mean one executor per machine. It works with one executor per machine, but often fails with 8 executors per machines. The machines have plenty of memory.

**Comment by randomjohnnyh** [comment]
I also occasionally receive a similar memcpy error on RabitFinalize:

```
Stack: [0x00007f0627831000,0x00007f0627932000],  sp=0x00007f062792f578,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libc.so.6+0x89835]  memcpy+0x35

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 21199  ml.dmlc.xgboost4j.java.XGBoostJNI.RabitFinalize()I (0 bytes) @ 0x00007f0696347fbe [0x00007f0696347f80+0x3e]
J 21198 C1 ml.dmlc.xgboost4j.java.Rabit.shutdown()V (7 bytes) @ 0x00007f069645773c [0x00007f06964576e0+0x5c]
j  ml.dmlc.xgboost4j.scala.spark.XGBoost$$anonfun$buildDistributedBoosters$1.apply(Lscala/collection/Iterator;)Lscala/collection/Iterator;+214
j  ml.dmlc.xgboost4j.scala.spark.XGBoost$$anonfun$buildDistributedBoosters$1.apply(Ljava/lang/Object;)Ljava/lang/Object;+5
J 8582 C1 org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$20.apply(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; (17 bytes) @ 0x00007f06964cccfc [0x00007f06964cc740+0x5bc]
J 19827 C2 org.apache.spark.rdd.MapPartitionsRDD.compute(Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;)Lscala/collection/Iterator; (36 bytes) @ 0x00007f0695d6c6fc [0x00007f0695d6c4a0+0x25c]
J 19432 C1 org.apache.spark.CacheManager.getOrCompute(Lorg/apache/spark/rdd/RDD;Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;Lorg/apache/spark/storage/StorageLevel;)Lscala/collection/Iterator; (433 bytes) @ 0x00007f0696d5321c [0x00007f0696d51ba0+0x167c]
J 11344 C2 org.apache.spark.rdd.RDD.iterator(Lorg/apache/spark/Partition;Lorg/apache/spark/TaskContext;)Lscala/collection/Iterator; (59 bytes) @ 0x00007f06955e8e0c [0x00007f06955e8d00+0x10c]
J 12364 C1 org.apache.spark.scheduler.ResultTask.runTask(Lorg/apache/spark/TaskContext;)Ljava/lang/Object; (170 bytes) @ 0x00007f0695249104 [0x00007f0695247500+0x1c04]
J 11144 C1 org.apache.spark.scheduler.Task.run(JILorg/apache/spark/metrics/MetricsSystem;)Lscala/Tuple2; (257 bytes) @ 0x00007f0696ac1b34 [0x00007f0696ac0120+0x1a14]
J 9757 C1 org.apache.spark.executor.Executor$TaskRunner.run()V (1951 bytes) @ 0x00007f0696810cc4 [0x00007f069680e5c0+0x2704]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
J 9575 C1 java.lang.Thread.run()V (17 bytes) @ 0x00007f069661c704 [0x00007f069661c5c0+0x144]
v  ~StubRoutines::call_stub
```

The fact that this only occurs at init/finalize suggests a networking problem - what could it be? Would switching to YARN help?

**Comment by randomjohnnyh** [comment]
After adding a bunch of fprintf's, I think the issue is in allreduce_base.cc, under
```
if (!tracker.Connect(utils::SockAddr(tracker_uri.c_str(), tracker_port))) {
```

I've tried to increase somaxconn but it does not work. Any ideas?

**Comment by randomjohnnyh** [comment]
Narrowed it further down to
```
memcpy(&addr.sin_addr, hp->h_addr_list[0], hp->h_length);
```
But this is incredibly strange: it works for 7/8 threads and fails for the last, even though all 8 are querying the same string.

**Comment by randomjohnnyh** [comment]
Addressed by #1409
